sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -765,14 +669,42 @@ int gcoap_req_init(coap_pkt_t *pdu, uint8_t *buf, size_t len,
     }
 }
 
+/*
+ * Assumes pdu.payload_len attribute was reduced in gcoap_xxx_init() to
+ * ensure enough space in PDU buffer to write Content-Format option and
+ * payload marker here.
+ */
 ssize_t gcoap_finish(coap_pkt_t *pdu, size_t payload_len, unsigned format)
 {
-    /* reconstruct full PDU buffer length */
-    size_t len = pdu->payload_len + (pdu->payload - (uint8_t *)pdu->hdr);
+    if (pdu->options_len && payload_len) {
```

Comment:
Actually, we make this friendlier by adding another condition here. We won't check the assertion if the user specifies there is not a Content-Format option (`format` is `COAP_FORMAT_NONE`).

Reply:
sounds valid, so +1

Side note: I think this check should be written as a single logical expression inside the `assert()` (intentional bad formatting...:
```c
assert( !(pdu->options_len) ||
        !(payload_len) || 
        (pdu->options[pdu->options_len-1].opt_num < COAP_OPT_CONTENT_FORMAT));
```
hope I did the logic correctly here...

Reply:
OK, I'll rework the assert. Your construct is more idiomatic although as you say the logic is a little harder to follow.